subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -44,58 +44,108 @@ type HivedAlgorithm struct {
 	algorithmLock sync.RWMutex
 }
 
-// NewHivedAlgorithm initializes a HivedAlgorithm from the config file
+// NewHivedAlgorithm initializes a HivedAlgorithm from the config file.
 func NewHivedAlgorithm(sConfig *api.Config) *HivedAlgorithm {
-	pcl, gpuNums, gpuTypeToChain, cellLevelToType, nonReservedVcl, reservedVcl, reservedPc := ParseConfig(sConfig)
+	fullPcl, freePcl, vcQuotas, nonReservedFullVcl, nonReservedFreeVcl, reservedVcl, reservedPc,
+		gpuNums, gpuTypeToChain, cellLevelToType := ParseConfig(sConfig)
+
 	h := &HivedAlgorithm{
-		vcSchedulers:            make(map[api.VirtualClusterName]intraVCScheduler),
+		vcSchedulers:            map[api.VirtualClusterName]intraVCScheduler{},
 		opportunisticSchedulers: map[CellChain]*topologyAwareScheduler{},
-		fullCellList:            pcl,
-		freeCellList:            make(map[CellChain]ChainCellList),
+		fullCellList:            fullPcl,
+		freeCellList:            freePcl,
+		vcFreeCellNum:           vcQuotas,
+		totalLeftCellNum:        map[CellChain]map[CellLevel]int32{},
+		badFreeCellNum:          map[CellChain]map[CellLevel]int32{},
+		badNodes:                common.NewSet(),
 		chains:                  gpuTypeToChain,
 		cellTypes:               cellLevelToType,
-		allocatedAffinityGroups: make(map[string]*AlgoAffinityGroup),
-		reservedCells:           reservedPc,
+		allocatedAffinityGroups: map[string]*AlgoAffinityGroup{},
+		apiClusterStatus: api.ClusterStatus{
+			PhysicalCluster: api.PhysicalClusterStatus{},
+			VirtualClusters: map[api.VirtualClusterName]api.VirtualClusterStatus{},
+		},
 	}
-	for vc := range nonReservedVcl {
+	for vcName := range nonReservedFullVcl {
 		// TODO: Support per-VC configurable intra VC scheduling algo.
-		h.vcSchedulers[vc] = newDefaultIntraVCScheduler(nonReservedVcl[vc], reservedVcl[vc], gpuNums)
+		h.vcSchedulers[vcName] = newDefaultIntraVCScheduler(
+			nonReservedFullVcl[vcName], nonReservedFreeVcl[vcName], reservedVcl[vcName], gpuNums)
 	}
 	for chain, ccl := range h.fullCellList {
 		h.opportunisticSchedulers[chain] = NewTopologyAwareScheduler(ccl, gpuNums[chain], false, true)
 	}
 	h.validateInitialAssignment()
-	h.initFreeCellList()
-	h.initReservations()
+	h.initAPIClusterStatus()
+	h.initReservations(reservedPc)
+	h.initBadNodes()
 	return h
 }
 
 func (h *HivedAlgorithm) AddNode(node *core.Node) {
-	// TODO
+	h.algorithmLock.Lock()
+	defer h.algorithmLock.Unlock()
+
+	if !internal.IsNodeHealthy(node) {
+		// adding a bad node
+		h.setBadNode(node.Name)
+	} else {
+		// possibly a bad node comes back again
+		h.setHealthyNode(node.Name)
+	}
 }
 
 func (h *HivedAlgorithm) UpdateNode(oldNode, newNode *core.Node) {
-	// TODO
+	h.algorithmLock.Lock()
+	defer h.algorithmLock.Unlock()
+
+	if oldHealthy := internal.IsNodeHealthy(oldNode); oldHealthy != internal.IsNodeHealthy(newNode) {
+		if oldHealthy {
+			h.setHealthyNode(oldNode.Name)
+			h.setBadNode(newNode.Name)
+		} else {
+			h.setBadNode(oldNode.Name)
+			h.setHealthyNode(newNode.Name)
+		}
```

Comment:
Seems it can be simple as:
  if oldHealthy {
   h.setBadNode(newNode.Name)
  } else {
   h.setHealthyNode(newNode.Name)
  }